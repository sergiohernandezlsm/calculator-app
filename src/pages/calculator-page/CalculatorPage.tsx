import { useState, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import ProductTable from '../../components/product-table';
import { ProductTypes, FieldTypes, StateTypes, ProductTableTypes } from '../../types';
import FormInput from '../../components/form-input';
import data from '../../services/data.json';
import { useEffect } from 'react';

const CalculatorPage = () => {
  const [principal, setPrincipal] = useState(0);
  const [revolvingCreditData, setRevolvingCreditData] = useState([]);
  const [businessCreditData, setBusinessCreditData] = useState([]);
  const [totalBusinessCredit, setTotalBusinessCredict] = useState({ totalPrincipal: 0, totalInterest: 0, totalRepayment: 0 });
  const [totalRevolvingCredit, setTotalRevolvingCredit] = useState({ totalPrincipal: 0, totalInterest: 0, totalRepayment: 0 });
  const amount = useSelector((state: StateTypes) => state.amount);
  const duration = useSelector((state: StateTypes) => state.duration);
  const revolvingCredit = useSelector((state: StateTypes) => state.revolvingCredit);
  const buninessCredit = useSelector((state: StateTypes) => state.buninessCredit);
  const dispatch = useDispatch();

  const calculatorHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.name === 'amount') {
      dispatch({ type: 'amount', payload: +e.target.value });
    }
    if (e.target.name === 'duration') {
      dispatch({ type: 'duration', payload: +e.target.value });
    }
    if (e.target.name === 'revolvingCredit') {
      dispatch({ type: 'revolvingCredit', payload: +e.target.value });
    }
    if (e.target.name === 'buninessCredit') {
      dispatch({ type: 'buninessCredit', payload: +e.target.value });
    }
  }

  const calculateRevolvingCredit = useCallback(() => {
    let initialCalculation: any = [] as ProductTableTypes[];
    for (let i = 0; i < duration; i++) {
      initialCalculation.push({
        principal: principal,
        interest: i !== 0 ? ((amount - (principal * i)) / 100) * revolvingCredit.value : (amount / 100) * revolvingCredit.value,
        totalRepayment: (i !== 0 ? ((amount - (principal * i)) / 100) * revolvingCredit.value : (amount / 100) * revolvingCredit.value) + principal,
      })
    }
    setRevolvingCreditData(initialCalculation);
    setTotalRevolvingCredit({
      totalPrincipal: initialCalculation.map((x: ProductTableTypes) => x.principal).reduce((total: number, amount: number) => {
        return total + amount
      }, 0),
      totalInterest: initialCalculation.map((x: ProductTableTypes) => x.interest).reduce((total: number, amount: number) => {
        return total + amount
      }, 0),
      totalRepayment: initialCalculation.map((x: ProductTableTypes) => x.totalRepayment).reduce((total: number, amount: number) => {
        return total + amount
      }, 0)
    });
  }, [amount, duration, principal, revolvingCredit.value])

  const calculateBusinessCredit = useCallback(() => {
    let initialCalculation: any = [] as ProductTableTypes[];
    for (let i = 0; i < duration; i++) {
      initialCalculation.push({
        principal: principal,
        interest: i !== 0 ? ((amount - (principal * i)) / 100) * buninessCredit.value : (amount / 100) * buninessCredit.value,
        totalRepayment: (i !== 0 ? ((amount - (principal * i)) / 100) * buninessCredit.value : (amount / 100) * buninessCredit.value) + principal,
      })
    }
    setBusinessCreditData(initialCalculation);
    setTotalBusinessCredict({
      totalPrincipal: initialCalculation.map((x: ProductTableTypes) => x.principal).reduce((total: number, amount: number) => {
        return total + amount
      }, 0),
      totalInterest: initialCalculation.map((x: ProductTableTypes) => x.interest).reduce((total: number, amount: number) => {
        return total + amount
      }, 0),
      totalRepayment: initialCalculation.map((x: ProductTableTypes) => x.totalRepayment).reduce((total: number, amount: number) => {
        return total + amount
      }, 0)
    });
  }, [amount, buninessCredit.value, duration, principal])

  useEffect(() => {
    if (amount !== 0 && duration !== 0) {
      setPrincipal(amount / duration);
    }
    if (principal !== 0 && revolvingCredit.value !== 0) {
      calculateRevolvingCredit();
    }
    if (principal !== 0 && buninessCredit.value !== 0) {
      calculateBusinessCredit();
    }
  }, [amount, buninessCredit.value, calculateBusinessCredit, calculateRevolvingCredit, duration, principal, revolvingCredit.value])

  const newGeneralData = data.productsData.map(x => {
    if (x.name === 'revolvingCredit') {
      return {
        ...x,
        productData: revolvingCreditData,
        totalRow: totalRevolvingCredit
      }
    }
    if (x.name === 'buninessCredit') {
      return {
        ...x,
        productData: businessCreditData,
        totalRow: totalBusinessCredit
      }
    }
    return {
      ...x,
      productData: [],
      totalRow: {}
    }
  });

  return (
    <Container>
      <Row>
        {data.formFields.map((field: FieldTypes, index: number) => {
          return (
            <FormInput
              key={`key-${index}`}
              placeholder={field.placeholder}
              name={field.name}
              text={field.text}
              label={field.label}
              min={10}
              max={100}
              calculatorHandler={calculatorHandler}
            />
          )
        })}
      </Row>
      <Row>
        {newGeneralData.map((loan: ProductTypes, index: number) => {
          return (
            <ProductTable
              productField={loan.formData}
              calculatorHandler={calculatorHandler}
              title={loan.title}
              key={`key-${index}`}
              id={index}
              productData={loan.productData}
              totalRow={loan?.totalRow}
            />
          );
        })}
      </Row>
    </Container>
  )
};

export default CalculatorPage;
