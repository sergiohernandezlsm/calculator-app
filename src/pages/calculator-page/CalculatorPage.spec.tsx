import { useSelector, useDispatch } from 'react-redux';
import { shallow, mount } from 'enzyme';
import CalculatorPage from './CalculatorPage';
import faker from 'faker';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const mockDefaultData = {
  revolvingCredit: {
    value: faker.datatype.number(),
    name: 'revolvingCredit'
  },
  businessCredit: {
    value: faker.datatype.number(),
    name: 'businessCredit'
  },
  amount: 10000,
  duration: 6,
  principal: 100,
  credits: [
    {
      title: faker.lorem.words(3),
      creditName: "revolvingCredit",
      formData: {}
    },
    {
      title: faker.lorem.words(3),
      creditName: "businessCredit",
      formData: {}
    }
  ],
  requestForm: [{
    name: "amount",
    placeholder: faker.lorem.word(),
    text: "(in Â£)",
    label: faker.lorem.word()
  },
  {
    name: "duration",
    placeholder: faker.lorem.word(),
    text: "(in months)",
    label: faker.lorem.word()
  }]
}

describe('Calculator Page Component', () => {

  beforeEach(() => {
    (useDispatch as jest.Mock).mockImplementation(() => mockDispatch);
    (useSelector as jest.Mock).mockImplementation(f => f(mockDefaultData));
    jest.clearAllMocks();
  });

  it('renders <CalculatorPage /> component', () => {
    const wrapper = shallow(<CalculatorPage />);
    expect(wrapper).toMatchSnapshot();
  });

  it('to have been call dispatch', () => {
    const wrapper = mount(<CalculatorPage />);
    const formInput = wrapper.find('[name="amount"]');
    const input = formInput.find('.inputStyles').at(0);
    input.simulate('change', 50);
    expect(mockDispatch).toHaveBeenCalledTimes(4);
  });

  it('renders with defult data', () => {
    const mockNewData = {
      ...mockDefaultData,
      credits: [{}]
    };
    (useSelector as jest.Mock).mockImplementation(f => f(mockNewData));
    const wrapperWithoutCreditsData = shallow(<CalculatorPage />);
    expect(wrapperWithoutCreditsData).toMatchSnapshot();
  });
});