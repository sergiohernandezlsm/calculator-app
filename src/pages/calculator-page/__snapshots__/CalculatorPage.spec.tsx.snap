// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Page Component renders <CalculatorPage /> component 1`] = `ShallowWrapper {}`;

exports[`Calculator Page Component renders with defult data 1`] = `ShallowWrapper {}`;
